### PR TITLE
Feature/cb2 11709 test: remove nested integration tests

### PR DESCRIPTION
**Pull request by FurkanErdogan-BJSS**
## Description

- Remove nested integration tests
- Move unit test execution to pre-commit
- Fix vulnerabilities
- Fix ci issues regarding package versions
- Remove --coverage from integration test execution in workflow to avoid heap memory issue
- Update jest configs to categorise unit and integration tests
- Add vscode jest settings to enable executing integration tests from vscode
- Add jest-serial-runner to allow sequential tests execution for integration tests

Related issue: [11709](https://dvsa.atlassian.net/browse/CB2-11709)

## Before submitting (or marking as "ready for review")

- [x] Does the pull request title follow the [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/) specification?
- [x] Have you performed a self-review of the code
- [ ] Have you have added tests that prove the fix or feature is effective and working
- [x] Did you make sure to update any documentation relating to this change?

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -1,13 +1,32 @@
...
+      testMatch: ["<rootDir>/tests/**/*.unitTest.ts"],
+      preset: "ts-jest",
+      testEnvironment: "node",
+      roots: ["<rootDir>/tests"],
+      setupFiles: ["jest-plugin-context/setup"],
+      moduleFileExtensions: ["js", "ts"],
+      testResultsProcessor: "jest-sonar-reporter",
+      transform: {
+        "^.+\\.tsx?$": "ts-jest",
+      },
```
Comment (Nat-McCabe-BJSS) [inline jest.config.js:15]: Do we need `runner: "jest-serial-runner"` in here as well?

File: package.json [hunk 2]
```diff
@@ -18,12 +18,13 @@
...
+    "test-all": "cross-env SQS_BUCKET=local AWS_REGION=local BRANCH=local USE_CONTAINERIZED_DATABASE=1 AWS_XRAY_CONTEXT_MISSING=LOG_ERROR jest --runInBand --coverage --collectCoverageFrom=\"src/!(models)/**/!(*conversion).{js,ts}\"",
+    "test": "cross-env SQS_BUCKET=local AWS_REGION=local BRANCH=local AWS_XRAY_CONTEXT_MISSING=LOG_ERROR jest --selectProjects UNIT --runInBand --coverage --collectCoverageFrom=\"src/!(models)/**/!(*conversion).{js,ts}\"",
+    "test-i": "cross-env SQS_BUCKET=local AWS_REGION=local USE_CONTAINERIZED_DATABASE=0 BRANCH=local jest --selectProjects INTEGRATION --runInBand",
+    "test-local-i": "cross-env SQS_BUCKET=local AWS_REGION=local BRANCH=local USE_CONTAINERIZED_DATABASE=1 jest --selectProjects INTEGRATION --runInBand --coverage --collectCoverageFrom=\"src/!(models)/**/!(*conversion).{js,ts}\"",
     "lint": "tslint src/**/*.ts tests/**/*.ts -q",
     "prettier": "prettier --write ./**/*.{js,ts}",
-    "prepush": "npm test && npm run build && npm run test-local-i",
+    "prepush": "npm run build && npm run test-local-i",
```
Comment (Nat-McCabe-BJSS) [inline package.json:27]: ~~Dumb question, why are we removing the unit tests from this? Makes sense to me to run `test-all`~~ 

Sorry I've just seen the reasoning is in the description of the github description, ignore this

**Comment by Nat-McCabe-BJSS** [comment]
Nice work, happy to approve once we look into the single line suggestion I made :)

**Review by Nat-McCabe-BJSS — DISMISSED** [review]

File: jest.config.js (hunk 1 above)
Comment (FurkanErdogan-BJSS) [reply jest.config.js:15]: Talked to Nat already but responding here for reference;
Unit tests are fast so we don't necessarily need to run them sequentially.
Serial runner was added to potentially fix the heap memory issue when running integration tests. It turned out the issue was caused by the --coverage argument. So this serial runner even for integration tests might not be that necessary anymore.

**Review by meys-bjss — COMMENTED** [review]
looks good to me

File: .vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (naathanbrown) [inline .vscode/settings.json]: this folder needs to be in .gitignore and this file removed

File: jest.config.js [hunk 4]
```diff
@@ -1,13 +1,32 @@
...
+      displayName: "INTEGRATION",
+      testMatch: ["<rootDir>/tests/**/*.intTest.ts"],
+      preset: "ts-jest",
+      testEnvironment: "node",
+      roots: ["<rootDir>/tests"],
+      setupFiles: ["jest-plugin-context/setup"],
+      moduleFileExtensions: ["js", "ts"],
+      testResultsProcessor: "jest-sonar-reporter",
+      transform: {
+        "^.+\\.tsx?$": "ts-jest",
```
Comment (naathanbrown) [inline jest.config.js]: So I was tired when I talked about this! this needs to be come .ts and all the files renamed need to go back to .ts...

.tsx is a react file

**Comment by naathanbrown** [comment]
Would be amazing to run these in the pipeline pre merge to make sure it's happy with the changes. https://github.com/dvsa/cvs-devops/actions -> this repo has all the actions needed to build a feature env and make sure that you're working. Think you just need to run the new-code-cut job

**Comment by FurkanErdogan-BJSS** [comment]
> Would be amazing to run these in the pipeline pre merge to make sure it's happy with the changes. https://github.com/dvsa/cvs-devops/actions -> this repo has all the actions needed to build a feature env and make sure that you're working. Think you just need to run the new-code-cut job

Started this run: https://github.com/dvsa/cvs-devops/actions/runs/8878463350 
not 100% sure if I did it correctly though.

**Review by naathanbrown — DISMISSED** [review]

**Review by Nat-McCabe-BJSS — APPROVED** [review]
Looks good, nice one!

File: jest.config.js (hunk 1 above)
Comment (FurkanErdogan-BJSS) [reply jest.config.js:15]: Resolving this as the changes are approved by @Nat-McCabe-BJSS

File: .vscode/settings.json (hunk 3 above)
Comment (FurkanErdogan-BJSS) [reply .vscode/settings.json]: Resolving this as the changes are approved by @naathanbrown

File: jest.config.js (hunk 4 above)
Comment (FurkanErdogan-BJSS) [reply jest.config.js]: Resolving this as the changes are approved by @naathanbrown

**Review by DanielFry-bjss — APPROVED** [review]

**Pull request merged**